### PR TITLE
add A40 batch inference; deal with OOM in compute_metrics

### PR DESCRIPTION
**Pull request by xzzWZY**

### CODE REVIEW

File: scripts/compute_system_metrics.py [hunk 1]
```diff
@@ -21,9 +21,12 @@ def main(data_dir: str, out_file: str) -> None:
         for benchmark_file in glob(f"{data_dir}/{model_name}/benchmark_batch_*.json"):
             batch_size = int(benchmark_file.split("_")[-1][:-5])
             df = pd.read_json(benchmark_file)
-            out_csv.writerow(
-                [model_name.replace("--", "/"), str(batch_size)] + df[metrics].mean().to_list(),
-            )
+            if len(df) < 100:
```
Comment (jaywonchung) [inline scripts/compute_system_metrics.py]: Please calculate the exact number of items. Should it be around `len(share_gpt_dataset) // batch_size`?

**Review by jaywonchung — CHANGES_REQUESTED** [review]
I don't think `app.py` will start with `OOM` in the energy column. It's not a floating point number. Please remove all lines with OOM and make sure `python app.py` shows results that make sense.

**Comment by jaywonchung** [comment]
@xzzWZY I just pushed https://github.com/ml-energy/leaderboard/commit/aaadf663f3d035d76a9c290bc34c85fc7ffde717 so that you can run `python app.py` without having to specify the Colosseum controller address. You can do the following to get the new `app.py`:

```sh
git checkout master
git pull
git checkout a40_benchmark
git rebase master
git push -f
```

File: scripts/compute_system_metrics.py (hunk 1 above)
Comment (xzzWZY) [reply scripts/compute_system_metrics.py]: yes it would be, but not the case for llama-2-7b-chat, since I mentioned before that llama-2-7b-chat will generate wierd nan in logits and I'll just skip that batch for llama-2-7b-chat

Comment (jaywonchung) [reply scripts/compute_system_metrics.py]: Ah makes sense. I'll resolve.

**Comment by xzzWZY** [comment]
> I don't think `app.py` will start with `OOM` in the energy column. It's not a floating point number. Please remove all lines with OOM and make sure `python app.py` shows results that make sense.

remove OOM in A40_{task}_benchmark.csv, and verify the output
for the original csv with 'OOM', it doesn't in df returned from _read_table() as you say, it's not an float

**Comment by jaywonchung** [comment]
I also did a quick check and apparently only the A40 GPU tabs work and not V100 and A100. Maybe because those CSV files don't have the `batch_size` column.

**Comment by xzzWZY** [comment]
yep I test with deleting "A100" and "V100" in schema.yaml

**Comment by jaywonchung** [comment]
That's not a meaningful test though. We're not going to delete V100 and A100 in the real leaderboard just to add batch size for A40.

**Comment by xzzWZY** [comment]
maybe we can just add batch_size 1 for each row in A100 and V100 currently

**Comment by jaywonchung** [comment]
Of course.

**Comment by jaywonchung** [comment]
You have to let me know whenever you push something. I don't get notified.

`response_length` decreases with higher batch size. I actually don't understand why. Shouldn't the response length numbers be roughly equal regardless of batch size?

**Comment by jaywonchung** [comment]
I sanity-checked the total number of tokens in each output (`sum(tokenizer.encode(batch["output"][i]) for i in range(batch_size))`) divided by the total number of outputs (`sum(len(batch["output"]) for batch in benchmark_json`) mostly matches the `response_length` number. Sometimes they are different by a few tokens but I think those differences come from `output.strip()`.

**Comment by jaywonchung** [comment]
I really hope this isn't a bug in `benchmark.py` but rather me not getting something

**Comment by xzzWZY** [comment]
I find for batch inference, it prefer to generate more empty output

**Comment by jaywonchung** [comment]
I'm not sure why batching should change model output

**Comment by xzzWZY** [comment]
> I sanity-checked the total number of tokens in each output (`sum(tokenizer.encode(batch["output"][i]) for i in range(batch_size))`) divided by the total number of outputs (`sum(len(batch["output"]) for batch in benchmark_json`) mostly matches the `response_length` number. Sometimes they are different by a few tokens but I think those differences come from `output.strip()`.

that's because I suppose all the stop_str has only one token. I do so for easy implementation at beginning, and I check that all model we benchmark has very short "stop_str", and I didn't fix this bug in an easy way.

It's a bug of benchmark.py, but it should have equal inference on batch inference for different batch size.

**Comment by jaywonchung** [comment]
I think I found the bug. Please run this yourself too and investigate further, and if you agree that this is the source of the bug, post a PR to fix `benchmark.py`. I think we should have `attention_mask`.

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer


model = AutoModelForCausalLM.from_pretrained("Salesforce/xgen-7b-8k-inst", device_map="cuda")

tok = AutoTokenizer.from_pretrained("Salesforce/xgen-7b-8k-inst", trust_remote_code=True)
tok.pad_token = tok.eos_token
tok.padding_side = "left"

# Different sequences have different lengths, and that's why we pad from the left for generation.
inputs = ["Hi how are you doing?", "Can you come up with five to ten ideas to celebrate a seven year old boy?"]
inputs_encoded = tok(inputs, padding=True)
input_ids = torch.as_tensor(inputs_encoded.input_ids, device="cuda")
# The beginning of shorter sequences (padded with `pad_token`) has attention_mask value 0.
attention_masks = torch.as_tensor(inputs_encoded.attention_mask, device="cuda")

torch.inference_mode().__enter__()
out_with_mask1 = model(input_ids, attention_mask=attention_masks, use_cache=True)
out_with_mask2 = model(input_ids, attention_mask=attention_masks, use_cache=True)
out_without_mask1 = model(input_ids, use_cache=True)
out_without_mask2 = model(input_ids, use_cache=True)

# Just to make sure repeating `model()` always yields the same result
assert torch.allclose(out_with_mask1.logits, out_with_mask2.logits)
assert torch.allclose(out_without_mask1.logits, out_without_mask2.logits)

# This assertion fails. With and without attention mask are different.
assert torch.allclose(out_with_mask1.logits, out_without_mask1.logits), "attention_mask matters"
```

**Comment by xzzWZY** [comment]
I agree that add attntion_mask would be better
previously I inherit from fastchat code without using attention_mask, but I find now that adding attention_mask may mask the padding, which make the batch inference make more sense

and I also run your test code and find the output of adding attention_mask is different, I'm current in a course lab and I'll modify "generate" a response later to check if adding attention_mask solve the problem of decreasing response length later

**Comment by jaywonchung** [comment]
I think it's incorrect without `attention_mask`. And I think because the left-padded EOS tokens were not properly hidden from the model, models were immediately outputting EOS on the first iteration of generation following the pattern. With larger batch sizes more and more sequences will have gotten EOS tokens in the beginning and more would have output an EOS token immediately, explaining the decrease in average response length.

**Comment by xzzWZY** [comment]
that make sense

I notice "batch inference prefers to generate more empty output" previously, but didn't go deep about it... I'm sorry I didn't discuss with you about this observation, causing a lot of waste of time and resource(

**Comment by jaywonchung** [comment]
We all learn by making mistakes. After all these, I'm pretty sure both you and I understand very deeply about LLM text generation. It's a gain.

I think this PR should be closed instead of merging, because it holds inaccurate data. Let's update `scripts/compute_system_metrics.py` in the following PR that fixes `scripts/benchmark.py`.

Another unrelated point, but I'm sure there are (or at least there will be) cases where `stop_str` is longer than one token. Please fix that along the way.
